hotshot-state-prover/src/service.rs

```diff
@@ -306,16 +308,34 @@ pub async fn read_contract_state(
 pub async fn submit_state_and_proof(
     proof: Proof,
     public_input: PublicInput,
-    provider: Url,
-    key: SigningKey,
-    light_client_address: Address,
+    contract: &LightClient<SignerWallet>,
 ) -> Result<(), ProverError> {
-    let contract = prepare_contract(provider, key, light_client_address).await?;
-
     // prepare the input the contract call and the tx itself
     let proof: ParsedPlonkProof = proof.into();
     let new_state: ParsedLightClientState = public_input.into();
-    let tx = contract.new_finalized_state(new_state.into(), proof.into());
+
+    let mut tx = contract.new_finalized_state(new_state.into(), proof.into());
+
+    // only use gas oracle for mainnet
+    if contract.client_ref().get_chainid().await?.as_u64() == 1 {
+        let gas_oracle = BlockNative::new(None).category(GasCategory::SafeLow);
+        match gas_oracle.estimate_eip1559_fees().await {
+            Ok((max_fee, priority_fee)) => {
+                if let TypedTransaction::Eip1559(inner) = &mut tx.tx {
+                    inner.max_fee_per_gas = Some(max_fee);
+                    inner.max_priority_fee_per_gas = Some(priority_fee);
+                    tracing::info!(
+                        "Setting maxFeePerGas: {}; maxPriorityFeePerGas to: {}",
+                        max_fee,
+                        priority_fee
+                    );
+                }
+            }
+            Err(e) => {
+                tracing::warn!("!! BlockNative Price Oracle failed: {}", e);
```

Comment:
by default, there's a tx filler. which will have some heuristic of choosing gas price, but will be less frugal (i.e. instead of choosing `SafeLow`, it will choose `Standard`)